### ISSUE TITLE
Allow custom outputDirectory like other Maven Plugins

### ISSUE THREAD

**Issue by brianoliver** [jacoco/jacoco#395]
### Steps to reproduce

JaCoCo version: (all)
Operating system: (all) 
Tool integration: Maven

### Expected behaviour
As a developer I expect the JaCoCo Maven Plugin will work in similar ways to other standard Maven Plugins.   For example, being able to customize the output directory for instrumented classes.

Being forced to use the standard output directory prevents simple interoperability with other plugins.  It also forces projects to adopt a directory structure that may not be suitable.

eg: Anyone wishing to use the Assembly or Shade Plugins while generating OSGi Manifests (in different folders) may not be able to use JaCoCo with in the same build lifecycle.   This leads to developers having to create additional modules (for code coverage generation) and/or new build lifecycles.

Providing an option to use an alternate output directory solves this challenge.

For example:
```xml
     <plugin>
        <groupId>org.jacoco</groupId>
        <artifactId>jacoco-maven-plugin</artifactId>
        <executions>
          <execution>
            <id>instrument</id>
            <phase>package</phase>
            <goals>
              <goal>instrument</goal>
            </goals>
            <configuration>
              <outputDirectory>${instrumented.classes.path}</outputDirectory>
            </configuration>
          </execution>
      </plugin>
```

### Actual behaviour
Sadly the JaCoCo Maven plugin doesn't allow customized a output directory (unlike the Ant Task) when instrumenting classes.

**Comment by marchof** [comment]
Here is how the goal currently works:

- The original classes from target/classes are copied to target/generated-classes/jacoco
- The classes in target/classes get replaced with instrumented versions

Which aspects do you want to have configurable?
- The backup folder?
- The target folder?

The idea is to use the goal restore-instrumented-classes to restore the original classes from the backup copy before any assembly is created, see our Example: http://eclemma.org/jacoco/trunk/doc/examples/build/pom-offline.xml

Can you please detail why this doesn't work for you?

**Comment by brianoliver** [comment]
Imagine a library/artifact/module (jar) that once produced, has to go through a large, multi-stage integration, compatibility and security test across many servers.  In such a situation, one needs to produce an instrumented jar that can be "installed" in one or more repositories, which are then consumed by multiple components (and servers) over the course of a week or more, through an extensive pipeline, of functional and integration tests, which ultimately "run" the artifact across in excess of 100,000 processes (jvm start/stops).

Note: While one may argue that code-coverage shouldn't be used in situations like this, it's very clear that without running the artifact(s) in these types of environments we miss out on a lot of code-coverage analytics and reporting.

Here's the fundamental challenges.

1. On the fly instrumentation, which takes around 20 seconds for one package, is out of the question.   Adding 100,000 * 20 seconds to the build time is not only a waste of time, it's an excessive waste of CPU cycles.  eg: over 500 hrs would be lost, just for instrumentation.  This means offline is the only reasonable solution.  

2. Ultimately we'd like to produce two artifacts as part of module.  The "regular" artifact and another that is classified as being for "code-coverage".   eg: produce my-artifact-1.0.0.jar and my-artifact-1.0.0-jacoco.jar.   By doing this we could run our existing application and tests with either jar, simply by adding a classifier to the dependencies (which could be parameterized).   


At the moment it's very difficult with the way that the JaCoCo Maven Plugin is designed to instrument in one **hard-coded** target/classes location.

3. The classes that we'd like instrumented are not in the target/classes folder.   They are generated by an intermediate compiler and need to be placed outside of the usual target/classes folder, otherwise other Maven tools erroneously pick them up.  

Hence we need to instrument the original classes that are in a "custom" location, not target/classes.   

While we could copy the 10,000+ classes into the target/classes folder and then back again, that too is a waste of time and CPU. 

4. The only plugin we use that doesn't allow customizing where classes are manipulated / stored etc, is the JaCoCo plugin.

5. Ultimately we'd like to customize both the target and backup folder.  

Personally I think it's the wrong way around.   The target/classes (original aka: target) folder should never be touched.   By default, the instrumented classes should be placed elsewhere, eg: target/generated-classes.   There should be an option to "replace" the target/classes with the instrumented classes, but not as the default, in which case the original classes could be backed up (if required) into target/classes-original.

Then it's very clear as to what classes are where at all times.

BTW: I'll help in anyway to resolve it.  ie: I can setup a pull request for you review / integrate.

**Comment by marchof** [comment]
Thanks for clarification!

So you need three locations configurable:

- source classes
- backup target
- instrumented classes target

In-place instrumentation and restore afterwards has been implemented to minimize the impact on existing tests within the plug-in. If you need to publish instrumented artifacts this is of course another valid use case.

**Comment by brianoliver** [comment]
Yes please!   Configuring all three locations would make our lives much easier, though we'd probably only use one of these.

I've modified the latest master line (in a local branch) to support changing the sources location.   I'm hapy to submit a pull request, just for you to review / provide feedback on if you like?  It works great in our build system.

My last wish was that we didn't need to keep the "originals" around for reporting.   It seems that to generate coverage reports from *.exec files we need to provide the "original classes".   This means we have to produce both the instrumented and non-instrumented jars in our build.   Of course it's possible but if there was some way (dreaming I know) that we could generate the coverage reports using the instrumented classes (jar) that would simplify things even further.

Many thanks!

**Comment by Godin** [comment]
@brianoliver I believe that you can already generate coverage reports using non-instrumented classes + exec , however such report will have granularity down to classes/methods, but not down to lines.

**Comment by marchof** [comment]
@brianoliver Unfortunatelly we need the original class files to analyze the control flow and match with the execution data. There is no way to "uninstrument" the classes.

**Comment by brianoliver** [comment]
@marchof Thanks for the explanation.   Perhaps one day it will be possible.   eg: when a jar can support multiple versions of the same classes (coming soon).

**Comment by Godin** [comment]
As a note: I didn't found similar options in `cobertura` and `clover` Maven plugins, i.e. seems that they able to perform only replacement in-place, as we do nowadays.

In any case - IMO either we want to do such improvement, either we should close this ticket. So @marchof what do you think, if @brianoliver will share his implementation of this for the review? If this improvement will be transparent for existing configurations, i.e. won't break compatibility for our users, then I personally don't see big problems to integrate such improvement.

**Comment by marchof** [comment]
@Godin Well we could make at least the location for the backup target and the instrumented classes configurable -- using the current values as defaults. So this shouldn't break existing builds.

**Comment by davidlinse** [comment]
Any news on this ?

**Comment by brianoliver** [comment]
Sorry all.  I'll create a pull request for this today so all can review my workaround.

**Comment by brianoliver** [comment]
Created a pull request for the prototype.
